### PR TITLE
Misc perf

### PR DESCRIPTION
**Pull request by Therzok**
TODO: Figure out whether GetImportFiles/ImportFile can be optimized in any way.

Not too big gains, but quick ones, for around 30MB on solution load. Containing code still allocates around 140MB (was 170).

### CODE REVIEW

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 1]
```diff
@@ -1115,9 +1117,20 @@ void Evaluate (ProjectInfo project, MSBuildEvaluationContext context, MSBuildImp
...
-					if (import.Project.IndexOf ("$(" + prop.Property + ")", StringComparison.OrdinalIgnoreCase) == -1)
+				var importSearchPaths = context.GetProjectImportSearchPaths ();
+				for (int i = 0; i < importSearchPaths.Count; ++i) {
+					var prop = importSearchPaths [i];
+
+					int propertyStart = import.Project.IndexOf ("$(", StringComparison.Ordinal);
+					if (propertyStart == -1)
+						continue;
+					int propertyNameStart = import.Project.IndexOf (prop.Property, propertyStart, StringComparison.OrdinalIgnoreCase);
+					if (propertyNameStart == -1)
```
[the remark this answers is not in the document]
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Oh, forgot that, good catch!

**Comment by Therzok** [comment]
`ProjectLoadSaveTests` sets an EnvironmentVariable at the beginning of the test. If any other test hits evaluation before that happens.

Added do not merge for the test until I profile it again to see perf stats.

**Comment by Therzok** [comment]
The other thing that could be done is invalidate the env var cache via the tests, but that feels hacky.

**Comment by Therzok** [comment]
This is good to go. Just verified, performance is good.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 2]
```diff
@@ -1115,9 +1117,22 @@ void Evaluate (ProjectInfo project, MSBuildEvaluationContext context, MSBuildImp
 			// In that case, look in fallback search paths
 
 			if (keepSearching) {
-				foreach (var prop in context.GetProjectImportSearchPaths ()) {
-					if (import.Project.IndexOf ("$(" + prop.Property + ")", StringComparison.OrdinalIgnoreCase) == -1)
+				var importSearchPaths = context.GetProjectImportSearchPaths ();
+				for (int i = 0; i < importSearchPaths.Count; ++i) {
+					var prop = importSearchPaths [i];
+
+					int propertyStart = import.Project.IndexOf ("$(", StringComparison.Ordinal);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: This causes a perf regression and correctness regression. Fixing it up.

**Comment by Therzok** [comment]
Old timings:
```
100.00%   Evaluate  •  34,967 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(Object, Boolean)
  99.94%   EvaluateProject  •  34,945 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext)
    98.49%   EvaluateObjects  •  34,439 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateObjects(ProjectInfo, MSBuildEvaluationContext, IEnumerable[MSBuildObject], Boolean)
      45.88%   Evaluate  •  16,044 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, Boolean)
        ►22.73%   ImportFile  •  7,947 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.ImportFile(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, String, String)
        ►18.44%   GetImportFiles  •  6,447 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.GetImportFiles(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, String, String, String&, Boolean&)
         1.79%   IndexOf  •  626 ms  •  System.String.IndexOf(String, StringComparison)
        ►1.16%   EvaluateProject  •  405 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext, Boolean)
        ►1.13%   Append  •  396 ms  •  MonoDevelop.Projects.ConditionedPropertyCollection.Append(ConditionedPropertyCollection)
         0.13%   Concat  •  45 ms  •  System.String.Concat(String, String, String)
        ►0.11%   MSBuildEvaluationContext..ctor  •  37 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext..ctor(MSBuildEvaluationContext)
         0.10%   Add  •  34 ms  •  System.Collections.Generic.List`1.Add(T)
         0.06%   GetEnumerator  •  21 ms  •  System.Collections.Generic.List`1.GetEnumerator
         0.05%   get_Property  •  16 ms  •  MonoDevelop.Projects.Extensions.ImportSearchPathExtensionNode.get_Property
        ►0.04%   GetImportedProjects  •  13 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.GetImportedProjects(ProjectInfo, MSBuildImport)
         0.03%   GetProjectImportSearchPaths  •  11 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.GetProjectImportSearchPaths
      ►35.20%   Evaluate  •  12,308 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildPropertyGroup)
      ►6.68%   Evaluate  •  2,334 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImportGroup, Boolean)
      ►5.44%   Evaluate  •  1,901 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildTarget)
      ►3.34%   Evaluate  •  1,169 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildItemGroup)
      ►1.31%   Evaluate  •  456 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildChoose, Boolean)
       0.35%   MoveNext  •  122 ms  •  System.Linq.Enumerable+<OfTypeIterator>d__93`1.MoveNext
    ►0.97%   InitEvaluation  •  339 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.InitEvaluation(MSBuildProject)
    ►0.34%   EvaluateWithItems  •  121 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.EvaluateWithItems(String, List[MSBuildItemEvaluated])
     0.08%   Exists  •  28 ms  •  System.IO.File.Exists(String)
     0.05%   TryGetValue  •  18 ms  •  System.Collections.Generic.Dictionary`2.TryGetValue(TKey, TValue&)
  ►0.03%   SetPropertyValue  •  11 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
   0.03%   get_Current  •  11 ms  •  System.Collections.Generic.Dictionary+Enumerator`2.get_Current
```

New timings:
```
100.00%   Evaluate  •  19,672 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(Object, Boolean)
  99.47%   EvaluateProject  •  19,568 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext)
    98.40%   EvaluateObjects  •  19,357 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateObjects(ProjectInfo, MSBuildEvaluationContext, IEnumerable[MSBuildObject], Boolean)
      ►42.77%   Evaluate  •  8,414 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildPropertyGroup)
      36.56%   Evaluate  •  7,193 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, Boolean)
        ►21.15%   GetImportFiles  •  4,160 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.GetImportFiles(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, String, String, String&, Boolean&)
        ►10.91%   ImportFile  •  2,146 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.ImportFile(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, String, String)
        ►1.37%   Append  •  270 ms  •  MonoDevelop.Projects.ConditionedPropertyCollection.Append(ConditionedPropertyCollection)
         1.05%   IndexOf  •  206 ms  •  System.String.IndexOf(String, Int32, StringComparison)
        ►0.91%   EvaluateProject  •  180 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext, Boolean)
         0.39%   Add  •  78 ms  •  System.Collections.Generic.List`1.Add(T)
         0.12%   IndexOf  •  24 ms  •  System.String.IndexOf(String, StringComparison)
        ►0.08%   GetImportedProjects  •  17 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.GetImportedProjects(ProjectInfo, MSBuildImport)
         0.08%   MSBuildEvaluationContext..ctor  •  15 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext..ctor(MSBuildEvaluationContext)
         0.04%   set_IsImported  •  7 ms  •  MonoDevelop.Projects.MSBuild.MSBuildItemEvaluated.set_IsImported(Boolean)
      ►6.23%   Evaluate  •  1,225 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildItemGroup)
      ►5.11%   Evaluate  •  1,006 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildTarget)
      ►4.00%   Evaluate  •  787 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImportGroup, Boolean)
      ►2.06%   Evaluate  •  405 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildChoose, Boolean)
       0.86%   MoveNext  •  169 ms  •  System.Linq.Enumerable+<OfTypeIterator>d__93`1.MoveNext
    ►0.91%   InitEvaluation  •  178 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.InitEvaluation(MSBuildProject)
     0.16%   Exists  •  32 ms  •  System.IO.File.Exists(String)
  ►0.41%   MSBuildEvaluationContext..ctor  •  81 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext..ctor
  ►0.06%   SetPropertyValue  •  11 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
```

**Comment by Therzok** [comment]
Current bottleneck is a new lock contention on DefaultMSBuildEngine.conditionCache. To investigate perf gains for locking per condition file or having an immutable dictionary.

**Comment by Therzok** [comment]
Load before this PR
oldload

Load after this PR
newload

**Comment by Therzok** [comment]
Caching the condition factor evaluation:
```
100.00%   Evaluate  •  17,896 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(Object, Boolean)
  99.69%   EvaluateProject  •  17,840 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext)
    97.85%   EvaluateObjects  •  17,510 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateObjects(ProjectInfo, MSBuildEvaluationContext, IEnumerable[MSBuildObject], Boolean)
      ►41.95%   Evaluate  •  7,507 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, Boolean)
      31.46%   Evaluate  •  5,630 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildPropertyGroup)
        28.16%   Evaluate  •  5,040 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildProperty)
          21.54%   SafeParseAndEvaluate  •  3,855 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.SafeParseAndEvaluate(ProjectInfo, MSBuildEvaluationContext, String, Boolean, String)
            7.53%   TryEvaluateToBool  •  1,347 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionRelationalExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
              ►6.55%   TryEvaluateToBool  •  1,172 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFactorExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
              ►0.41%   TryEvaluateToVersion  •  74 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFactorExpression.TryEvaluateToVersion(IExpressionContext, Version&)
              ►0.27%   TryEvaluateToString  •  49 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFactorExpression.TryEvaluateToString(IExpressionContext, String&)
              ►0.22%   TryEvaluateToNumber  •  39 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFactorExpression.TryEvaluateToNumber(IExpressionContext, Single&)
            ►4.76%   TryEvaluateToBool  •  852 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionAndExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
             4.53%   Enter  •  811 ms  •  System.Threading.Monitor.Enter(Object, Boolean&)
            ►1.64%   TryEvaluateToBool  •  293 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionOrExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
            ►1.19%   CollectConditionProperties  •  213 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionAndExpression.CollectConditionProperties(ConditionedPropertyCollection)
            ►0.62%   CollectConditionProperties  •  111 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionRelationalExpression.CollectConditionProperties(ConditionedPropertyCollection)
             0.34%   TryGetValue  •  60 ms  •  System.Collections.Generic.Dictionary`2.TryGetValue(TKey, TValue&)
            ►0.17%   CollectConditionProperties  •  30 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionOrExpression.CollectConditionProperties(ConditionedPropertyCollection)
             0.12%   set_CustomFullDirectoryName  •  21 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.set_CustomFullDirectoryName(String)
            ►0.07%   ParseCondition  •  13 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionParser.ParseCondition(String)
            ►0.07%   TryEvaluateToBool  •  13 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionNotExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
            ►0.07%   TryEvaluateToBool  •  12 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFunctionExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
          ►4.74%   Evaluate  •  849 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.Evaluate(String, Boolean&)
          ►0.85%   SetPropertyValue  •  152 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
          ►0.74%   StoreProperty  •  132 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.StoreProperty(ProjectInfo, String, String, String)
           0.15%   get_Name  •  27 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyCore.get_Name
        ►1.74%   SafeParseAndEvaluate  •  311 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.SafeParseAndEvaluate(ProjectInfo, MSBuildEvaluationContext, String, Boolean, String)
        ►0.88%   MoveNext  •  157 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyGroup+<GetProperties>d__35.MoveNext
        ►0.68%   GetProperties  •  122 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyGroup.GetProperties
      ►10.26%   Evaluate  •  1,836 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildTarget)
      ►8.29%   Evaluate  •  1,483 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildItemGroup)
      ►3.90%   Evaluate  •  698 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImportGroup, Boolean)
      ►0.84%   Evaluate  •  150 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildChoose, Boolean)
       0.63%   MoveNext  •  112 ms  •  System.Linq.Enumerable+<OfTypeIterator>d__93`1.MoveNext
    ►0.88%   InitEvaluation  •  158 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.InitEvaluation(MSBuildProject)
    ►0.63%   EvaluateWithItems  •  113 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.EvaluateWithItems(String, List[MSBuildItemEvaluated])
     0.17%   Exists  •  30 ms  •  System.IO.File.Exists(String)
     0.09%   GetEnumerator  •  16 ms  •  System.Collections.Generic.List`1.GetEnumerator
    ►0.07%   SetPropertyValue  •  13 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
  ►0.16%   MSBuildEvaluationContext..ctor  •  29 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext..ctor
```

**Comment by Therzok** [comment]
And finally, using an immutable dictionary instead of locking on a dictionary:
```
100.00%   Evaluate  •  8,422 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(Object, Boolean)
  98.63%   EvaluateProject  •  8,306 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateProject(ProjectInfo, MSBuildEvaluationContext)
    92.73%   EvaluateObjects  •  7,810 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.EvaluateObjects(ProjectInfo, MSBuildEvaluationContext, IEnumerable[MSBuildObject], Boolean)
      ►39.73%   Evaluate  •  3,346 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImport, Boolean)
      34.55%   Evaluate  •  2,910 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildPropertyGroup)
        30.10%   Evaluate  •  2,535 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildProperty)
          18.02%   SafeParseAndEvaluate  •  1,517 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.SafeParseAndEvaluate(ProjectInfo, MSBuildEvaluationContext, String, Boolean, String)
            ►9.21%   TryEvaluateToBool  •  775 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionRelationalExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
            ►4.25%   TryEvaluateToBool  •  358 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionAndExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
            ►1.18%   CollectConditionProperties  •  99 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionRelationalExpression.CollectConditionProperties(ConditionedPropertyCollection)
            ►0.80%   CollectConditionProperties  •  67 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionAndExpression.CollectConditionProperties(ConditionedPropertyCollection)
            ►0.73%   TryEvaluateToBool  •  61 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionOrExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
             0.72%   TryGetValue  •  60 ms  •  System.Collections.Immutable.ImmutableDictionary`2.TryGetValue(TKey, TValue&)
            ►0.28%   CollectConditionProperties  •  24 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionOrExpression.CollectConditionProperties(ConditionedPropertyCollection)
            ►0.19%   TryEvaluateToBool  •  16 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionNotExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
            ►0.14%   ParseCondition  •  12 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionParser.ParseCondition(String)
            ►0.11%   TryEvaluateToBool  •  9 ms  •  MonoDevelop.Projects.MSBuild.Conditions.ConditionFunctionExpression.TryEvaluateToBool(IExpressionContext, Boolean&)
          ►9.17%   Evaluate  •  772 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.Evaluate(String, Boolean&)
          ►1.57%   StoreProperty  •  133 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.StoreProperty(ProjectInfo, String, String, String)
          ►0.77%   SetPropertyValue  •  65 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
           0.11%   get_Name  •  10 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyCore.get_Name
        ►2.06%   SafeParseAndEvaluate  •  173 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.SafeParseAndEvaluate(ProjectInfo, MSBuildEvaluationContext, String, Boolean, String)
        ►1.62%   MoveNext  •  136 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyGroup+<GetProperties>d__35.MoveNext
        ►0.61%   GetProperties  •  51 ms  •  MonoDevelop.Projects.MSBuild.MSBuildPropertyGroup.GetProperties
      ►6.80%   Evaluate  •  573 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildTarget)
      ►6.49%   Evaluate  •  547 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildItemGroup)
      ►2.60%   Evaluate  •  219 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildImportGroup, Boolean)
       1.44%   MoveNext  •  121 ms  •  System.Linq.Enumerable+<OfTypeIterator>d__93`1.MoveNext
       0.13%   Evaluate  •  11 ms  •  MonoDevelop.Projects.MSBuild.DefaultMSBuildEngine.Evaluate(ProjectInfo, MSBuildEvaluationContext, MSBuildChoose, Boolean)
    ►4.98%   InitEvaluation  •  420 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.InitEvaluation(MSBuildProject)
     0.80%   Exists  •  67 ms  •  System.IO.File.Exists(String)
    ►0.12%   SetPropertyValue  •  10 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext.SetPropertyValue(String, String)
  ►0.53%   MSBuildEvaluationContext..ctor  •  45 ms  •  MonoDevelop.Projects.MSBuild.MSBuildEvaluationContext..ctor
   0.14%   Clear  •  12 ms  •  System.Collections.Generic.Dictionary`2.Clear
```

**Comment by Therzok** [comment]
Gif for the latest version:
lastload

**Comment by Therzok** [comment]
Old:
latestold

New:
latestsmall

**Comment by Therzok** [comment]
Synchronized gifs.

New:
ezgif-2-f2b663393c

Old:
ezgif-2-8da0d95adb

**Comment by Therzok** [comment]
Removed condition factor evaluation caching for now. Need to figure out what cases should trigger invalidation inside an evaluation context.

**Comment by Therzok** [comment]
@monojenkins backport d15-7

cc @slluis

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 3]
```diff
@@ -174,11 +191,28 @@ public override object CreateProjectInstance (object project)
 			return pi;
 		}
 
-		public override void DisposeProjectInstance (object projectInstance)
```
Comment (slluis) [inline main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Why this change? Why do we need the project instance as IDisposable? I don't see why this change is necessary

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: See the change here:
https://github.com/mono/monodevelop/pull/4335/files#diff-46ff2760f1591d9bd1d02f42550c7d62L87

Other bits of code keep a reference to the imported projects until the new one is loaded.

i.e. https://github.com/mono/monodevelop/blob/bcebeb6e8ffabf3dcc1b98123a2a8fbb2b008eb4/main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs#L202

Comment (slluis) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: But you can do the same by storing projectInstance in a temp var, and calling DisposeProjectInstance after evaluating.

Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Okay, yeah, that's simpler! 👍

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 4]
```diff
@@ -130,38 +132,53 @@ public override void UnloadProject (object project)
...
 				}
-				LogBeginProjectFileLoad (context, fileName);
-				MSBuildProject p = new MSBuildProject (EngineManager);
-				p.Load (fileName, new MSBuildXmlReader { ForEvaluation = true });
-				loadedProjects [fileName] = new LoadedProjectInfo { Project = p, LastWriteTime = File.GetLastWriteTime (fileName) };
-				LogEndProjectFileLoad (context);
-				return p;
+			}
+
+			lock (pi.LockObject) {
```
[the remark this answers is not in the document]
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Ah, this is a leftover. I used to have the IO code in-between the locks.

**Comment by Therzok** [comment]
TODO: add unit test asserting we only load imports once.

File: main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs [hunk 5]
```diff
@@ -130,38 +132,53 @@ public override void UnloadProject (object project)
...
-					var lastWriteTime = File.GetLastWriteTime (fileName);
-					if (pi.LastWriteTime != lastWriteTime) {
-						pi.LastWriteTime = lastWriteTime;
-						pi.Project.Load (fileName, new MSBuildXmlReader { ForEvaluation = true });
-					}
-					return pi.Project;
+				if (!loadedProjects.TryGetValue (fileName, out pi)) {
+					loadedProjects [fileName] = pi = new LoadedProjectInfo ();
+				}
+				pi.ReferenceCount++;
```
[the remark this answers is not in the document]
Comment (Therzok) [reply main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/DefaultMSBuildEngine.cs]: Seems like if I change this to be right, we end up double evaluating imports again. Will fix and add unit test

**Comment by Therzok** [comment]
Updated the PR. Now it should be OK, it has unit tests.
